### ISSUE TITLE
Argument matcher anyXxx() (i.e. anyString(), anyList()) should not match nulls

### ISSUE THREAD

**Issue by alexo** [mockito/mockito#134]
This is a bug I'm seeing in 1.10.8 version (older version has the same issue - tested with 1.9.0).

Given:
```java
Function<Object, Integer> function = Mockito.mock(Function.class);
when(function.apply(Mockito.anyString())).thenReturn(1);
Integer result = function.apply(2);
```
Expected behavior:
```java
result == null;
```

Actual behavior:
```java
result == 1;
```

Note that the function is called with an integer (not a string), and still the mocked function return the value which it should return only when a string is passed. The same works when using anyBoolean() or any other methof from any* family.

**Comment by szczepiq** [comment]
Included in 2.0.0-beta. Thanks for help!

**Comment by bric3** [comment]
For reference purpose see related work in #510 as well
